iceoryx_dust/vocabulary/include/iox/detail/span_iterator.hpp

```diff
@@ -0,0 +1,223 @@
+    {
+        assert(m_begin && m_end);
+        assert(m_begin <= m_current && m_current < m_end);
+        return m_current;
+    }
+    constexpr span_iterator& operator++() noexcept
+    {
+        assert(m_begin && m_current && m_end);
+        assert(m_current < m_end);
+
+        ++m_current;
+        return *this;
+    }
+
+    // Rule DCL21-CPP is deprecated
+    // NOLINTNEXTLINE(cert-dcl21-cpp)
+    constexpr span_iterator operator++(int) noexcept
+    {
+        span_iterator ret = *this;
+        ++*this;
+        return ret;
+    }
+
+    constexpr span_iterator& operator--() noexcept
+    {
+        assert(m_begin && m_end);
+        assert(m_begin < m_current);
+        --m_current;
+        return *this;
+    }
+
+    // Rule DCL21-CPP is deprecated
+    // NOLINTNEXTLINE(cert-dcl21-cpp)
+    constexpr span_iterator operator--(int) noexcept
+    {
+        span_iterator ret = *this;
+        --*this;
+        return ret;
+    }
+
+    constexpr span_iterator& operator+=(const difference_type n) noexcept
+    {
+        if (n != 0)
+        {
+            assert(m_begin && m_current && m_end);
+        }
+        if (n > 0)
+        {
+            assert(m_end - m_current >= n);
+        }
+        if (n < 0)
+        {
+            assert(m_current - m_begin >= -n);
+        }
+
+        m_current += n;
+        return *this;
+    }
+
+    constexpr span_iterator operator+(const difference_type n) const noexcept
+    {
+        span_iterator ret = *this;
+        ret += n;
+        return ret;
+    }
+
+    friend constexpr span_iterator operator+(const difference_type n, const span_iterator& rhs) noexcept
+    {
+        return rhs + n;
+    }
+
+    constexpr span_iterator& operator-=(const difference_type n) noexcept
+    {
+        if (n != 0)
+        {
+            assert(m_begin && m_current && m_end);
+        }
+        if (n > 0)
+        {
+            assert(m_current - m_begin >= n);
+        }
+        if (n < 0)
+        {
+            assert(m_end - m_current >= -n);
+        }
+        m_current -= n;
+        return *this;
+    }
+
+    constexpr span_iterator operator-(const difference_type n) const noexcept
+    {
+        span_iterator ret = *this;
+        ret -= n;
+        return ret;
+    }
+
+    template <class Type2, std::enable_if_t<std::is_same<std::remove_cv_t<Type2>, value_type>::value, int> = 0>
+    constexpr difference_type operator-(const span_iterator<Type2>& rhs) const noexcept
+    {
+        assert(m_begin == rhs.m_begin && m_end == rhs.m_end);
+        return m_current - rhs.m_current;
+    }
+
+    constexpr reference operator[](const difference_type n) const noexcept
+    {
+        return *(*this + n);
+    }
+
+    template <class Type2, std::enable_if_t<std::is_same<std::remove_cv_t<Type2>, value_type>::value, int> = 0>
+    constexpr bool operator==(const span_iterator<Type2>& rhs) const noexcept
+    {
+        assert(m_begin == rhs.m_begin && m_end == rhs.m_end);
```

Comment:
Thinking out loud. Can this be removed?

Reply:
We can do a lot of things :grin: However, in this case, I'd stick with the `assert` to make sure both `span_iterator`s refer to the same `span`, otherwise UB can happen.

Reply:
Sure, but since one line further down pointer are compared I thought it is superfluous. On the other hand, if both `m_current` are `nullptr` it would lead to a wrong result when the iterators accidentally refer to different `span`